src/Http/Paginators/OffsetRequestPaginator.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Saloon\Http\Paginators;
+
+use Closure;
+use GuzzleHttp\Promise\PromiseInterface;
+use Saloon\Contracts\Connector;
+use Saloon\Contracts\Request;
+use Saloon\Contracts\Response;
+
+// TODO 1: Look into serialising the Connector and original Request,
+//           to ensure that we can rebuild the paginator state without storing the entire multiverse.
+// TODO 2: Make it easier to extend the RequestPaginator. Preferably via callbacks, so we ideally don't even need separate classes.
+// TODO 3: Because both page-based and offset-based pagination just bumps the 'paging' number,
+//           would it make sense to do all pagination in _the_ RequestPaginator, but allow to specify the counter somehow?
+//         Maybe a callback that receives a copy of the original Request, as well as the latest Response (which has the corresponding latest Request and PendingRequest),
+//           and have that callback set the next 'page' on the new Request?
```

Comment:
That is very true. And, in theory, the PendingRequest should already have had middlewares applied, which would ideally help with the serialisation situation.

Although.. it'd probably be better to always run the middlewares, in case something is modified depending on page number, or something.

Reply:
Everything is run on the PendingRequest’s constructor, so once you have the instance, it has already been through all plugins and middleware and executed them, the pending request in this instance is completely ready to be sent and all the properties should be merged into it.